tcms/telemetry/static/telemetry/js/testing/status-matrix.js

```diff
@@ -0,0 +1,129 @@
+let table;
+const table_columns = [
+    {
+        data: null,
+        className: "table-view-pf-actions",
+        render: function (data, type, full, meta) {
+            const caseId = data.tc_id;
+
+            return `<span style="padding: 5px;">` +
+                `<a href="/case/${caseId}">TC-${caseId}: ${data.tc_summary}</a>` +
+                `</span>`;
+        },
+        sortable: false
+    },
+];
+
+$(document).ready(() => {
+    loadInitialProduct();
+    loadInitialTestPlans();
+
+    document.getElementById('id_product').onchange = () => {
+        update_version_select_from_product();
+        update_build_select_from_product(true);
+        updateTestPlanSelectFromProduct();
+
+        drawTable();
+    };
+
+    document.getElementById('id_version').onchange = drawTable;
+    document.getElementById('id_build').onchange = drawTable;
+    document.getElementById('id_test_plan').onchange = drawTable;
+
+    $('#id_after').on('dp.change', drawTable);
+    $('#id_before').on('dp.change', drawTable);
+
+    drawTable();
+});
+
+
+function drawTable() {
+    if (table) {
+        table.destroy();
+    }
+
+    const query = {};
+
+    const productId = $('#id_product').val();
+    if (productId) {
+        query['product_version__product_id'] = productId;
+    }
+
+    const versionId = $('#id_version').val();
+    if (versionId) {
+        query['product_version_id'] = versionId;
+    }
+
+    const buildId = $('#id_build').val();
+    if (buildId) {
+        query['build_id'] = buildId;
+    }
+
+    const testPlanId = $('#id_test_plan').val();
+    if (testPlanId) {
+        query['plan_id'] = testPlanId;
+    }
+
+    const dateBefore = $('#id_before');
+    if (dateBefore.val()) {
+        query['stop_date__lte'] = dateBefore.data('DateTimePicker').date().format('YYYY-MM-DD 23:59:59');
+    }
+
+    const dateAfter = $('#id_after');
+    if (dateAfter.val()) {
+        query['start_date__gte'] = dateAfter.data('DateTimePicker').date().format('YYYY-MM-DD 00:00:00');
+    }
+
+    jsonRPC('Testing.status_matrix', query, data => {
+
+        const testRuns = [];
+
+        data.forEach(element => {
+            element.executions.forEach(testRun => {
+                testRuns[testRun.run_id] = testRun.run__summary;
+            });
+        });
+
+        for (let i = 1; i < testRuns.length; i++) {
+            if (testRuns[i]) {
+                d3.select('.table > thead > tr').append('th').text(`TR-${i}`);
+            }
```

Comment:
> When I load the page without any filters I see 197 test runs and then, to the right I see a lot more columns without headers, some columns are empty while others have test executions inside of them.

you should not see that, and I need to find a way to reproduce it and fix it.

> This loop is wrong. Test runs have their unique IDs, not comming from the order of the elements in the data set. They also have summaries.

yes, they are not in order, but after this loop they will be. actually, after this loop, `testRuns` will be something like `["testSummary-1", "testSummary-2", <empty slot>, "testSummary-4"]`
<img width="745" alt="Screenshot 2019-06-14 at 23 13 02" src="https://user-images.githubusercontent.com/18421997/59535600-0d712d80-8efa-11e9-82cb-98faae28a873.png">
<img width="288" alt="Screenshot 2019-06-14 at 23 13 27" src="https://user-images.githubusercontent.com/18421997/59535601-0d712d80-8efa-11e9-8563-48091f65d315.png">

see? in the first picture I call `TestRun.objects.filter()` and visualize the results. In the second you can see this results aggregated into the `testRuns` array. the id of the test run matches its place in the array. and this is how it is supposed to be, because this is the way I fill the array:
`testRuns[execution.run_id] = execution.run__summary;`

Now that I think about it, I don't have a missing slot, and maybe if I do, I will run into some problems. Will try that.

Reply:
I changed to logic a bit. Please review